docs/articles/documentation/reference/test-api/testcontroller/README.md

```diff
@@ -25,11 +25,17 @@ test('My Test', async t => {
 });
 ```
 
-The test runner also uses the test controller to access the internal context required for the test API to operate. When you call [selectors](../../../guides/basic-guides/select-page-elements.md) and [client functions](../../../guides/basic-guides/obtain-client-side-info.md) from Node.js callbacks, pass the test controller explicitly, because the API cannot retrieve it from the context.
+The test runner uses the test controller to access the internal context required for the test API to operate. When you call [selectors](../../../guides/basic-guides/select-page-elements.md) and [client functions](../../../guides/basic-guides/obtain-client-side-info.md) from Node.js callbacks, pass the test controller explicitly, because the API cannot retrieve it from the context.
 
-## Implicit Test Controller Usage
+> If you explicitly call a test controller, do so in a stack trace that has the test body signature.
+>
+> Constructs like `setTimeout`/`setInterval`, Promises and native async functions create a clean stack frame. In it, test controller can't access a test body signature and resolve the test context.
+>
+> For this reason, use imported test controllers in a synchronous context.
```

Comment:
Your version is obviously clearer. But perhaps a `don't` instead of `can't`?

> **Do not** use an imported test controller in Promises, asynchronous functions, `setTimeout`/`setInterva`l callbacks, and other asynchronous routines. These routines do not preserve the original stack frame so that TestCafe cannot obtain the test controller instance.

Reply:
Yes, `do not` is probably fine but I usually prefer saying what you cannot do. Because `do not` usually implies some danger. Anyway, I'd rather ask Dirk.

Reply:
Yes, you can use 'Do not' here, but you are repeating it in the next sentence. It is always better to say what you should do instead of what you should not do.

Use [insert text here] instead of Promises, asynchronous functions, setTimeout/setInterval callbacks, and other asynchronous routines because these routines do not preserve the original stack frame. TestCafe cannot obtain the test controller instance from these routines.

OR

Promises, asynchronous functions, setTimeout/setInterval callbacks, and other asynchronous routines do not preserve the original stack frame. TestCafe cannot obtain the test controller instance from these routines.

Reply:
> Promises, asynchronous functions, setTimeout/setInterval callbacks, and other asynchronous routines do not preserve the original stack frame. TestCafe cannot obtain the test controller instance from these routines.